x/exchange/keeper/fulfillment.go

```diff
@@ -0,0 +1,375 @@
+				order.OrderId, rerr))
+			continue
+		}
+		if err := k.releaseHoldOnOrder(ctx, order); err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		var totalSellerFee sdk.Coins
+		if sellerSettlementFlatFee != nil && !sellerSettlementFlatFee.IsZero() {
+			totalSellerFee = totalSellerFee.Add(*sellerSettlementFlatFee)
+		}
+		if sellerRatioFee != nil && !sellerRatioFee.IsZero() {
+			totalSellerFee = totalSellerFee.Add(*sellerRatioFee)
+		}
+
+		i, seen := addrIndex[seller]
+		if !seen {
+			i = len(assetInputs)
+			addrIndex[seller] = i
+			assetInputs = append(assetInputs, banktypes.Input{Address: seller})
+			priceOutputs = append(priceOutputs, banktypes.Output{Address: seller})
+		}
+		assetInputs[i].Coins = assetInputs[i].Coins.Add(assets)
+		priceOutputs[i].Coins = priceOutputs[i].Coins.Add(price)
+
+		if !totalSellerFee.IsZero() {
+			j, known := feeAddrIndex[seller]
+			if !known {
+				j = len(feeInputs)
+				feeAddrIndex[seller] = j
+				feeInputs = append(feeInputs, banktypes.Input{Address: seller})
+			}
+			feeInputs[j].Coins = feeInputs[j].Coins.Add(totalSellerFee...)
+		}
+
+		filledOrders = append(filledOrders, exchange.NewFilledOrder(order, price, totalSellerFee))
+	}
+
+	if len(errs) > 0 {
+		return errors.Join(errs...)
+	}
+
+	totalAssets, totalPrice := sumAssetsAndPrice(orders)
+
+	if !exchange.CoinsEquals(totalPrice, sdk.Coins{msg.TotalPrice}) {
+		return fmt.Errorf("total price %q does not equal sum of ask order prices %q", msg.TotalPrice, totalPrice)
+	}
+
+	if !msg.BuyerSettlementFees.IsZero() {
+		feeInputs = append(feeInputs, banktypes.Input{Address: msg.Buyer, Coins: msg.BuyerSettlementFees})
+	}
+
+	assetOutputs := []banktypes.Output{{Address: msg.Buyer, Coins: totalAssets}}
+	priceInputs := []banktypes.Input{{Address: msg.Buyer, Coins: sdk.Coins{msg.TotalPrice}}}
+
+	if err := k.DoTransfer(ctx, assetInputs, assetOutputs); err != nil {
+		return fmt.Errorf("error transferring assets from sellers to buyer: %w", err)
+	}
+
+	if err := k.DoTransfer(ctx, priceInputs, priceOutputs); err != nil {
+		return fmt.Errorf("error transferring price from buyer to sellers: %w", err)
+	}
+
+	if err := k.CollectFees(ctx, marketID, feeInputs); err != nil {
+		return fmt.Errorf("error collecting settlement fees: %w", err)
+	}
+
+	// Collected last so that it's easier for a seller to fill asks without needing those funds first.
+	// Collected separately so it's not combined with the buyer settlement fees in the events.
+	if msg.BidOrderCreationFee != nil {
+		if err := k.CollectFee(ctx, marketID, buyer, sdk.Coins{*msg.BidOrderCreationFee}); err != nil {
+			return fmt.Errorf("error collecting create-ask fee %q: %w", msg.BidOrderCreationFee, err)
+		}
+	}
+
+	events := make([]proto.Message, len(orders))
+	for i, order := range filledOrders {
+		deleteAndDeIndexOrder(store, *order.GetOriginalOrder())
+		events[i] = exchange.NewEventOrderFilled(order)
+	}
+
+	k.emitEvents(ctx, events)
+	return nil
+}
```

Comment:
A lot of the logic here is very similar to what is in FillBids. Is there any logic that they can share just to reduce the amount of code?

Reply:
Possibly, but a lot of it is similar but different, e.g. what's an input and what's an output, or what type of ratio fee to use. It'd end up taking quite a bit of scaffolding to generalize.

Reply:
Also, I wrote `FillAsks` and `FillBids` before `SettleOrders` (and all the stuff needed by it). I can probably have these create a `Settlement` and have a function that processes everything for a settlement. Then all three would be doing the same thing, at least for all the transfers.

I'll work on that and get back to you.

Reply:
Update: I did overhaul `FillAsks` and `FillBids` to shorten them up quite a bit. They're still really similar, but just different enough that extracting more to generic helpers would have lots of "If asks, do this, if bids, do that". But I also split a `closeSettlement` method out of `SettleOrders` and reused that for all the transferring stuff with `FillAsks` and `FillBids` too.